feat: sqlite full text search

will7200 opened the pull request:
Closes #103 

This aims to be a drop in replacement for whoosh. with little to no API change for our customers.

1. Uses FTS4 as FTS5 requires some escaping of the search terms
2. Since the underlying search implementation strips out certain characters out during tokenization, these characters that might be of value to the customer are lost. Added functionality for the mask and filter operations to allows users to pass a callable to control their search better. Operations supported are those of sqlite3 column operators wrapped in pythonic operations.
  Examples:
   ```python
   db.search("lollipop", mask={"product": lambda col: col == 'ZEBRA-2'})
   db.search("lollipop", mask={"product": lambda col: col.like('%ZEBRA-2%')})
   db.search("lollipop", mask={"product": lambda col: col.in_(['ZEBRA-2', 'ZEBRA-1'])})
    ```

will7200 committed:
feat: sqlite full text search

cmutel commented on bw2data/search/indices.py:
```diff
@@ -51,19 +118,78 @@ def add_dataset(self, ds):
-        writer.commit()
+        with self.db.bind_ctx(MODELS):
+            for model in MODELS:
+                model.delete().where(model.code == ds["code"]).execute()
```
The `code` by itself isn't enough to guarantee uniqueness - the code is normally made up by hashing "core" data attributes like name, location, and unit. However, if I make a copy of a given database, these attributes won't change (and this is convenient - I can then easily find the corresponding dataset in the copy). So we need to search for both `code` and `database` attribute values - the combination of these two is guaranteed to be unique.

cmutel commented on bw2data/search/indices.py:
```diff
@@ -51,19 +118,78 @@ def add_dataset(self, ds):
-        index.delete_by_term("code", ds["code"])
+        with self.db.bind_ctx(MODELS):
+            for model in MODELS:
+                model.delete().where(model.code == ds["code"]).execute()
```
See above.

cmutel commented:
@will7200 Awesome! One small comment which is an easy fix, and then fix some test failures, and we should be GTG!

will7200 committed:
fix: specific database updates and star searches

will7200 committed:
fix: sqlite user-defined exception when searching

on later python versions > 3.11, the weights constructed for the search function are invalid as they include the primary identifier as an arg to `matchinfo`.

will7200 commented:
@cmutel fixed that and found the underlying causing those test failures. The last two remaining windows failures are due to:
```
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpgwrprhc7\\dxbqkfxeszgknljeyz.46382e12\\lci\\databases.db'
```
Which looks like has happened in past github runs.
